src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java

```diff
@@ -432,6 +441,76 @@ public void deleteTreesBelow(Path dir) throws IOException {
     }
   }
 
+  private java.nio.file.Path createJavaNioPath(Path path) throws FileNotFoundException {
+    final String pathStr = path.getPathString();
+    if (pathStr.chars().allMatch(c -> c < 128)) {
+      return Paths.get(pathStr);
+    }
+
+    // Paths returned from NativePosixFiles are Strings containing raw bytes from the filesystem,
+    // but Java's IO subsystem expects paths to be encoded in the current locale. We can avoid this
+    // assumption by converting the path to a URI, which permits percent-encoding of any octet.
+    final byte[] pathBytes = path.getPathString().getBytes(StandardCharsets.ISO_8859_1);
+    final StringBuilder builder = new StringBuilder(pathBytes.length * 3);
+    builder.append("file://");
+    for (byte b : pathBytes) {
+      // Keep a few known-safe bytes unescaped for debugging.
+      if (b == 0x2F // '/'
+          || b == 0x2D // '-'
+          || b == 0x5F // '_'
+          || b == 0x2E // '.'
+          || (b >= 0x30 && b <= 0x39) // '0' - '9'
+          || (b >= 0x41 && b <= 0x5A) // 'A' - 'Z'
+          || (b >= 0x61 && b <= 0x7A) // 'a' - 'z'
+          ) {
+        builder.append((char)b);
+      } else {
+        builder.append(String.format("%%%02X", b));
+      }
+    }
+
+    try {
+      return Paths.get(new URI(builder.toString()));
+    } catch (URISyntaxException e){
+      throw new FileNotFoundException(pathStr);
+    }
+  }
+
+  @Override
+  protected InputStream createFileInputStream(Path path)
+      throws FileNotFoundException, IOException {
+    final java.nio.file.Path nioPath = createJavaNioPath(path);
+    InputStream stream = null;
+    try {
+      stream = Files.newInputStream(nioPath, java.nio.file.StandardOpenOption.READ);
```

Comment:
After much debugging, I found two culprits of the breaking internal tests.

One is this line. I don't know why, but returning `new FileInputStream(nioPath.toFile())` works, but `Files.newInputStream` doesn't. Maybe something somewhere checks for `instanceof FileInputStream` -- which of course they shouldn't, but do so anyway -- and while I'd love to fix that, I'm not sure I can. So, would you be open to using FileInputStream here?

The other error is the logic in lines 490..511. Again, frustratingly, I have no idea why it's wrong. But considering that this logic -- no matter how reasonable -- doesn't serve the goal of this PR (i.e. to support non-ASCII characters), nor is it essential to it, would you be OK with removing it?

Reply:
> So, would you be open to using FileInputStream here?

Well, using `java.nio.Files` is a hard requirement to make Bazel's CI pass.

> The other error is the logic in lines 490..511. Again, frustratingly, I have no idea why it's wrong. But considering that this logic -- no matter how reasonable -- doesn't serve the goal of this PR (i.e. to support non-ASCII characters), nor is it essential to it, would you be OK with removing it?

I didn't add this logic for fun. The CI tests failed unless I added the seek-once (since tests assert trying to read a directory fails), and using the normal buffering stream caused a different CI failure.

There doesn't seem to be an easy way around using NIO. I've been digging into the JDK source, and Bazel's current behavior of forcing `LANG=en_US.ISO-8859-1` has a side effect on Linux of setting `sun.jnu.encoding` to ASCII-only mode, which prevents `java.io.File` from being able to represent bytes > 0x7F at all.

There are ... a lot of things to unpack here, and it's possible that I'll find something that can be made to work. But it would be much easier if the dependencies on `FileInputStream` could be identified and characterized.